### PR TITLE
Adopt lazy imports from Terra

### PR DESCRIPTION
**Pull request by manoelmarques**
### Summary
Adopt the API implemented in Qiskit/qiskit-terra#7525

### CODE REVIEW

File: qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py [hunk 1]
```diff
@@ -81,7 +81,8 @@ def driver_class_from_type(
                 raise MissingOptionalLibraryError(
                     libname=driver_type, name="ElectronicStructureDriverType"
                 )
-            driver_class.check_installed()
+            # instantiating the object will check if the driver is installed
+            _ = driver_class()
```
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:85]: This implicitly implies that:
1. all driver classes can be instantiated without any arguments
2. an instantiation of the class before we checked whether the requested method is supported

Is there a better way than this? Or are these implications fine?

Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:85]: Okay continuing to read on I think these are not really of concern. Leaving this here for other reviews to check and discuss :+1:

File: qiskit_nature/optionals.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    A lazy checker that a command-line tool is available. This just checks if the command is
+    not empty.
+    """
+
+    __slots__ = ("_command",)
+
+    def __init__(
+        self,
+        command: str,
+        *,
```
Comment (mrossinek) [inline qiskit_nature/optionals.py:33]: :slightly_smiling_face:

File: qiskit_nature/optionals.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    G16PROG,
+    name=GAUSSIAN_16_DESC,
+    msg="Please check that it is installed correctly",
+)
+
+PSI4 = "psi4"
+PSI4_DESC = "PSI4"
+PSI4PROG = which(PSI4)
+if PSI4PROG is None:
+    PSI4PROG = PSI4
```
Comment (mrossinek) [inline qiskit_nature/optionals.py:84]: Why is this needed? If `which` cannot find the command how would we be able to use it?

File: qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py [hunk 4]
```diff
@@ -81,7 +81,8 @@ def driver_class_from_type(
                 raise MissingOptionalLibraryError(
                     libname=driver_type, name="ElectronicStructureDriverType"
                 )
```
Comment (mrossinek) [inline qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:83]: Related to the logic above which is dealing with driver importing. Can this still occur? I only noticed because I assumed that during the move to lazy imports we can get rid of the Nature-specific `MissingOptionalLibraryError` entirely..

File: test/drivers/second_quantization/test_molecule_driver.py [hunk 5]
```diff
@@ -60,14 +62,16 @@ def test_invalid_kwarg(self):
...
     def test_driver(self, config):
         """test driver"""
         driver_type, hf_energy = config
         driver = ElectronicStructureMoleculeDriver(
             self._molecule, basis="sto3g", driver_type=driver_type
         )
-        driver_result = driver.run()
+        try:
+            driver_result = driver.run()
+        except MissingOptionalLibraryError as ex:
```
Comment (mrossinek) [inline test/drivers/second_quantization/test_molecule_driver.py:73]: I guess this relates to my comment above: how can this still occur?
Is it because we defer the lazy loading of pyscf into another class?

**Review by mrossinek — COMMENTED** [review]
In general this looks good to me :+1: I have some minor comments.

I did grep for `ImportError` on the repo and did find a few more occurrences though. Maybe those can be refactored, too?
```
qiskit_nature/drivers/second_quantization/gaussiand/gaussiandriver.py:257:        except ImportError as mnfe:
test/drivers/second_quantization/fcidumpd/test_driver_fcidump_dumper.py:138:        except ImportError:
test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py:320:        except ImportError as ex:
test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py:408:        except ImportError as ex:
test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py:434:        except ImportError as ex:
test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py:465:        except ImportError as ex:
test/test_readme_sample.py:41:        except ImportError as ex:
```
Grepping for `MissingOptionalLibraryError` gives the following:
```
qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py:216:            MissingOptionalLibraryError: Requires matplotlib.
qiskit_nature/drivers/second_quantization/vibrational_structure_molecule_driver.py:22:from qiskit.exceptions import MissingOptionalLibraryError
qiskit_nature/drivers/second_quantization/vibrational_structure_molecule_driver.py:54:            MissingOptionalLibraryError: Driver not installed.
qiskit_nature/drivers/second_quantization/vibrational_structure_molecule_driver.py:64:                    except MissingOptionalLibraryError as ex:
qiskit_nature/drivers/second_quantization/vibrational_structure_molecule_driver.py:73:                raise MissingOptionalLibraryError(
qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:22:from qiskit.exceptions import MissingOptionalLibraryError
qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:59:            MissingOptionalLibraryError: Driver not installed.
qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:72:                    except (MissingOptionalLibraryError, UnsupportMethodError) as ex:
qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:81:                raise MissingOptionalLibraryError(
test/drivers/second_quantization/test_molecule_driver.py:20:from qiskit.exceptions import MissingOptionalLibraryError
test/drivers/second_quantization/test_molecule_driver.py:73:        except MissingOptionalLibraryError as ex:
test/drivers/second_quantization/test_molecule_driver.py:92:                except MissingOptionalLibraryError as ex:
test/drivers/second_quantization/test_molecule_driver.py:156:        except MissingOptionalLibraryError as ex:
```

File: qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py (hunk 1 above)
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:85]: @mrossinek We used to have a method to check if driver was installed. I removed them favoring the Terra approach of using decorators. Since there is a decorator for the __init__ and all the driver classes that are in the list of the Molecule Driver support no arguments __init__ I just test it by instantiating. Otherwise, I can reinstate the previous `check_installed` method.

Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:85]: Yep makes sense :+1: When getting into later parts of the review I started to understand that a bit better, too. I left it here for completeness sake

File: qiskit_nature/optionals.py (hunk 3 above)
Comment (manoelmarques) [reply qiskit_nature/optionals.py:84]: @mrossinek Just because the LazySubprocessTester from Terra won't accept a None or empty command in the constructor and I still wanted to use its capabilities of actually running the command to see if it is installed.

File: qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py (hunk 4 above)
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:83]: Lazy imports is not about removing `MissingOptionalLibraryError` . It is just a delayed mechanism but stiil raises `MissingOptionalLibraryError` as was done before. That exception needs to be handled by other areas of code and it is raised here because the Molecule Driver has a search driver mechanism and will raise the same exception if no appropriate driver was found.

File: qiskit_nature/optionals.py (hunk 3 above)
Comment (mrossinek) [reply qiskit_nature/optionals.py:84]: Okay, then why is the same not being done in the gaussian case a few lines further up?

File: qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py (hunk 4 above)
Comment (mrossinek) [reply qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:83]: So the lazy import mechanism does not provide a utility which we could use for this use case? In that case I'm obviously fine with this :+1:

**Comment by manoelmarques** [comment]
> In general this looks good to me 👍 I have some minor comments.
> 
> I did grep for `ImportError` on the repo and did find a few more occurrences though. Maybe those can be refactored, too?
> 
> ```
> qiskit_nature/drivers/second_quantization/gaussiand/gaussiandriver.py:257:        except ImportError as mnfe:
> test/drivers/second_quantization/fcidumpd/test_driver_fcidump_dumper.py:138:        except ImportError:
> test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py:320:        except ImportError as ex:
> test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py:408:        except ImportError as ex:
> test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py:434:        except ImportError as ex:
> test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py:465:        except ImportError as ex:
> test/test_readme_sample.py:41:        except ImportError as ex:
> ```
> 
> Grepping for `MissingOptionalLibraryError` gives the following:
> 
> ```
> qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py:216:            MissingOptionalLibraryError: Requires matplotlib.
> qiskit_nature/drivers/second_quantization/vibrational_structure_molecule_driver.py:22:from qiskit.exceptions import MissingOptionalLibraryError
> qiskit_nature/drivers/second_quantization/vibrational_structure_molecule_driver.py:54:            MissingOptionalLibraryError: Driver not installed.
> qiskit_nature/drivers/second_quantization/vibrational_structure_molecule_driver.py:64:                    except MissingOptionalLibraryError as ex:
> qiskit_nature/drivers/second_quantization/vibrational_structure_molecule_driver.py:73:                raise MissingOptionalLibraryError(
> qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:22:from qiskit.exceptions import MissingOptionalLibraryError
> qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:59:            MissingOptionalLibraryError: Driver not installed.
> qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:72:                    except (MissingOptionalLibraryError, UnsupportMethodError) as ex:
> qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:81:                raise MissingOptionalLibraryError(
> test/drivers/second_quantization/test_molecule_driver.py:20:from qiskit.exceptions import MissingOptionalLibraryError
> test/drivers/second_quantization/test_molecule_driver.py:73:        except MissingOptionalLibraryError as ex:
> test/drivers/second_quantization/test_molecule_driver.py:92:                except MissingOptionalLibraryError as ex:
> test/drivers/second_quantization/test_molecule_driver.py:156:        except MissingOptionalLibraryError as ex:
> ```

@mrossinek I will review the ImportError to see if any can be refactored but the `MissingOptionalLibraryError` need to be checked and need to be added to the documentation as being raised. That exception didn't go away. It is still being thrown by the Lazy import framework. It just has been delayed at the point where it is used.

File: qiskit_nature/optionals.py (hunk 2 above)
Comment (manoelmarques) [reply qiskit_nature/optionals.py:33]: If it is about the formatting, it is `black`s formatter fault  :-)

Comment (mrossinek) [reply qiskit_nature/optionals.py:33]: No it's about [PEP 3102](https://www.python.org/dev/peps/pep-3102/) :+1:

File: qiskit_nature/optionals.py (hunk 3 above)
Comment (manoelmarques) [reply qiskit_nature/optionals.py:84]: Ahhh, because unfortunately Gaussian doesn't have a command line option like version or whatever that lets me just simply run it without side effect of actually running code. For this I created my own Lazy class derived from Terra that just checks if the command is None. Effectively just what was being done before: 
```
class NatureLazyCommandTester(LazyDependencyManager):
```

File: qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py (hunk 4 above)
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/electronic_structure_molecule_driver.py:83]: Yes, the Molecule Driver search for an appropriate driver based on an enumeration passed is our own custom implementation. There is nothing there for it but we raise the correct exception.

File: qiskit_nature/optionals.py (hunk 3 above)
Comment (mrossinek) [reply qiskit_nature/optionals.py:84]: Oh I was not careful. The first time I read it, it looked like you used `NatureLazyCommandTester` for Gaussian and Psi4. But I see now that the latter actually uses the Terra-provided one :+1:

File: test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py [hunk 6]
```diff
@@ -398,16 +394,11 @@ def test_uccsd_hf_qasm(self):
         self.assertAlmostEqual(result.total_energies[0], -1.138, places=2)
 
     @slow_test
```
Comment (mrossinek) [inline test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py:396]: Since these are slow tests, can we have the CI run them once for this PR to ensure they work as expected?

**Review by mrossinek — COMMENTED** [review]
Just one question otw this LGTM :slightly_smiling_face:

File: test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py (hunk 6 above)
Comment (manoelmarques) [reply test/algorithms/ground_state_solvers/test_groundstate_eigensolver.py:396]: ok, I added the `run_slow` label to this PR

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py [hunk 7]
```diff
@@ -75,9 +75,10 @@ def __init__(
         # If running from a jcf we need Gaussian™ 16 so check if we have a
         # valid install.
         if self._logfile is None:
-            check_valid()
+            _optionals.HAS_GAUSSIAN.require_now("GaussianForcesDriver __init__")
 
     @staticmethod
+    @_optionals.HAS_GAUSSIAN.require_in_call("GaussianForcesDriver from_molecule")
```
Comment (jakelishman) [inline qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py:81]: If you don't supply an argument in the decorator form, it should use the qualified name of the method automatically.  If you prefer, this line is equivalent to decorating with `@_optionals.HAS_GAUSSIAN.require_in_call` (i.e. without "calling" the decorator)

File: qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py [hunk 8]
```diff
@@ -51,13 +50,12 @@
 
 logger = logging.getLogger(__name__)
 
-try:
+if _optionals.HAS_PYQUANTE2:
```
Comment (jakelishman) [inline qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py:53]: I don't know if you mind, but this isn't a lazy test - the Boolean evaluation will be triggered as soon as this module is imported.

File: qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py [hunk 9]
```diff
@@ -243,11 +242,10 @@ def from_molecule(
         Returns:
             driver
         """
-        PyQuanteDriver.check_installed()
         PyQuanteDriver.check_method_supported(method)
         kwargs = {}
         if driver_kwargs:
-            args = inspect.getfullargspec(PyQuanteDriver.__init__).args
+            args = inspect.signature(PyQuanteDriver.__init__).parameters.keys()
```
Comment (jakelishman) [inline qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py:248]: I put some effort into trying to make sure the wrappers would propagate through the signature and documentation (`require_in_instance` does a bit of magic to wrap the `__init__` method).  Let me know if something's broken about that.

(`inspect.signature` is the recommended way, though)

File: qiskit_nature/optionals.py (hunk 3 above)
Comment (jakelishman) [reply qiskit_nature/optionals.py:84]: What functionality would you like in `LazySubprocessTester`?  I made it reject "empty" commands as a way of catching bugs because I couldn't think of any use-cases, but if you've got a use-case for empty commands, we could add it.

Here, I would note that calling `which` on import isn't lazy - if you wanted it for your Gaussian one, perhaps your derived class could run `shutil.which` to define its `_is_available`?  It depends how lazy you care about being, though.

**Review by jakelishman — COMMENTED** [review]
I hope the tools are useful - let us/me know if there's anything about the existing Terra objects that might want expanding.

File: qiskit_nature/optionals.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+from qiskit.utils import LazyDependencyManager, LazyImportTester, LazySubprocessTester
+
+
+class NatureLazyCommandTester(LazyDependencyManager):
+    """
+    A lazy checker that a command-line tool is available. This just checks if the command is
+    not empty.
+    """
+
+    __slots__ = ("_command",)
```
Comment (jakelishman) [inline qiskit_nature/optionals.py:28]: Also, you don't need to do this if you don't want, especially if you'd like to attach data to the objects.  To be honest, I probably shouldn't have done it in Terra to give you the choice of adding arbitrary data - from a Python perspective, it's a bad habit of mine that I over-use `__slots__` (it comes from me preferring stricter typing, where you can't add attributes to stuff).

File: qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py (hunk 7 above)
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/gaussiand/gaussian_forces_driver.py:81]: Oh that is helpful. Thanks for the tip. I will change it.

File: qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py (hunk 8 above)
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py:53]: Yes, I knew. We had those at the top because they are being used in different parts of the file and even in type annotations like:
```
 self._mol: pyquante_molecule = None
 self._bfs: basisset = None
``` 

I will see if it is possible to move them inside the methods.

File: qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py (hunk 9 above)
Comment (manoelmarques) [reply qiskit_nature/drivers/second_quantization/pyquanted/pyquantedriver.py:248]: Yes, the `getfullargspec` args started returning an empty array. I changed to `signature` which works.

File: qiskit_nature/optionals.py (hunk 10 above)
Comment (manoelmarques) [reply qiskit_nature/optionals.py:28]: Yes, since I was creating a lazy class I decided to stick to the idiom there but already had someone else asking me about it too :-) I will remove it.

**Comment by manoelmarques** [comment]
@jakelishman About the `LazySubprocessTester`: 

Nature has two drivers, PSI4 and Gaussian, that have their own commands. We use `which` as a way of figuring out if they are installed. In the PSI4 case, I used `LazySubprocessTester` because it is an improvement by running also the command. I don't use `LazySubprocessTester` for Gaussian though because it doesn't have any command option that would let it run without actually feeding chemistry input which gets complicated.

A subprocess tester that would be useful would be one where I construct it with a command string, but inside `_is_available` it runs `which` and if it returns None, is an error. Also it doesn't run the command automatically since there is no good way to run Gaussian. Maybe an additional flag in the `__init__` telling if it should run the command. 

I created a very simple subclass for Gaussian just to test None or not but I was thinking of doing what you suggested and put the `which` inside the `_is_available`. I will try this and you can take a look. If you think it is useful, it could be moved to Terra at some point.

**Review by mrossinek — APPROVED** [review]
LGTM! Thanks @manoelmarques!
